File: benches/comparison_async.cpp
#include <iostream>
#include <chrono>
#include <atomic>
#include <memory>
#include <type_traits>

// Windows COM ABI (stdcall) をエミュレート
#if defined(_WIN32)
#define STDMETHODCALLTYPE __stdcall
#else
#define STDMETHODCALLTYPE
#endif

// =========================================================
// 1. Manual COM Async Operation (The "Hardcore" C++ way)
// =========================================================

struct IAsyncOperation {
    virtual unsigned long STDMETHODCALLTYPE AddRef() = 0;
    virtual unsigned long STDMETHODCALLTYPE Release() = 0;
    virtual int STDMETHODCALLTYPE GetStatus(int* status) = 0;
    virtual int STDMETHODCALLTYPE GetResult(int* result) = 0;
};

class AsyncOperationCompleted : public IAsyncOperation {
    std::atomic<unsigned long> ref_count_;
    int result_;

public:
    explicit AsyncOperationCompleted(int result)
        : ref_count_(1), result_(result) {}

    unsigned long STDMETHODCALLTYPE AddRef() override {
        return ref_count_.fetch_add(1, std::memory_order_relaxed) + 1;
    }

    unsigned long STDMETHODCALLTYPE Release() override {
        unsigned long count = ref_count_.fetch_sub(1, std::memory_order_release) - 1;
        if (count == 0) {
            std::atomic_thread_fence(std::memory_order_acquire);
            delete this;
        }
        return count;
    }

    int STDMETHODCALLTYPE GetStatus(int* status) override {
        *status = 1; // Completed
        return 0; // S_OK
    }

    int STDMETHODCALLTYPE GetResult(int* result) override {
        *result = result_;
        return 0; // S_OK
    }
};

struct IMyAsyncOp {
    virtual unsigned long STDMETHODCALLTYPE AddRef() = 0;
    virtual unsigned long STDMETHODCALLTYPE Release() = 0;
    virtual IAsyncOperation* STDMETHODCALLTYPE GetStatusAsync() = 0;
};

class ManualComImpl : public IMyAsyncOp {
    std::atomic<unsigned long> ref_count_;

public:
    ManualComImpl() : ref_count_(1) {}

    unsigned long STDMETHODCALLTYPE AddRef() override {
        return ref_count_.fetch_add(1, std::memory_order_relaxed) + 1;
    }

    unsigned long STDMETHODCALLTYPE Release() override {
        unsigned long count = ref_count_.fetch_sub(1, std::memory_order_release) - 1;
        if (count == 0) {
            std::atomic_thread_fence(std::memory_order_acquire);
            delete this;
        }
        return count;
    }

    IAsyncOperation* STDMETHODCALLTYPE GetStatusAsync() override {
        return new AsyncOperationCompleted(1);
    }
};

// =========================================================
// 2. Modern C++ Implementation (Baseline)
// =========================================================

struct ReadyFuture {
    int value;
};

class ModernImpl {
public:
    int GetStatus() {
        return 1;
    }
};

// =========================================================
// Benchmarking Utilities
// =========================================================

template <class T>
void do_not_optimize(T&& datum) {
    using BaseType = typename std::remove_reference<T>::type;
    volatile BaseType* p = &datum;
    (void)p;
}

template <typename Func>
double measure_ns(const char* name, int iterations, Func func) {
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < iterations; ++i) {
        func();
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    double avg = static_cast<double>(duration) / iterations;

    std::cout << "[" << name << "] Average: " << avg << " ns" << std::endl;
    return avg;
}

int main() {
    const int ITERATIONS = 10000000; // 10M loops

    std::cout << "Running C++ Async Benchmarks (" << ITERATIONS << " iterations)..." << std::endl;
    std::cout << "-----------------------------------------------------" << std::endl;

    // Prepare a COM object.
    IMyAsyncOp* raw_obj = new ManualComImpl();

    // --- Allocation Benchmark ---

    // 1. Manual COM: async operation allocation
    measure_ns("Cpp_AsyncOp_New", ITERATIONS, [raw_obj]() {
        IAsyncOperation* op = raw_obj->GetStatusAsync();
        op->Release();
    });

    // 2. Modern C++: make_shared (ready state)
    measure_ns("Cpp_Make_Shared_Ready", ITERATIONS, []() {
        auto ptr = std::make_shared<ReadyFuture>(ReadyFuture{1});
        do_not_optimize(ptr);
    });

    // --- Dispatch Benchmark ---

    IAsyncOperation* op = raw_obj->GetStatusAsync();

    // 3. Virtual Method Call (COM ABI)
    measure_ns("Cpp_AsyncOp_GetStatus", ITERATIONS, [op]() {
        int status = 0;
        op->GetStatus(&status);
        do_not_optimize(status);
    });

    op->Release();

    // 4. Native direct call
    ModernImpl native;
    measure_ns("Cpp_Native_Call", ITERATIONS, [&native]() {
        do_not_optimize(native.GetStatus());
    });

    raw_obj->Release();

    return 0;
}

File: benches/comparison_async.rs
#[cfg(feature = "async-com")]
use kcom::{
    declare_com_interface, impl_com_interface, impl_com_object, pin_init, AsyncOperationRaw,
    AsyncStatus, ComObject, ComRc, GlobalAllocator, GUID, IUnknownVtbl, InitBox, InitBoxTrait,
    NTSTATUS,
};
#[cfg(feature = "async-com")]
use std::hint::black_box;
#[cfg(feature = "async-com")]
use std::time::Instant;

// =========================================================
// 1. Interface Definition (Async)
// =========================================================

#[cfg(feature = "async-com")]
declare_com_interface! {
    pub trait IMyAsyncOp: IUnknown {
        const IID: GUID = GUID {
            data1: 0x2222_3333, data2: 0x4444, data3: 0x5555,
            data4: [0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD],
        };
        async fn get_status(&self) -> i32;
    }
}

// =========================================================
// 2. kcom Implementation (Async)
// =========================================================

#[cfg(feature = "async-com")]
struct MyAsyncImpl;

#[cfg(feature = "async-com")]
unsafe impl IMyAsyncOp for MyAsyncImpl {
    type GetStatusFuture = core::future::Ready<i32>;
    type Allocator = GlobalAllocator;

    fn get_status(&self) -> impl InitBoxTrait<Self::GetStatusFuture, Self::Allocator, NTSTATUS> {
        InitBox::new(GlobalAllocator, pin_init!(core::future::ready(1)))
    }
}

#[cfg(feature = "async-com")]
impl_com_interface! {
    impl MyAsyncImpl: IMyAsyncOp {
        parent = IUnknownVtbl,
        methods = [get_status],
    }
}

#[cfg(feature = "async-com")]
impl_com_object!(MyAsyncImpl, IMyAsyncOpVtbl);

// =========================================================
// 3. Standard Rust Implementation (Baseline)
// =========================================================

#[cfg(feature = "async-com")]
struct ModernImpl;

#[cfg(feature = "async-com")]
impl ModernImpl {
    fn get_status(&self) -> i32 {
        1
    }
}

// =========================================================
// Benchmarking Utilities
// =========================================================

#[cfg(feature = "async-com")]
fn measure_ns<F>(name: &str, iterations: u64, mut func: F) -> f64
where
    F: FnMut(),
{
    let start = Instant::now();
    for _ in 0..iterations {
        func();
    }
    let duration = start.elapsed();
    let total_ns = duration.as_nanos() as f64;
    let avg = total_ns / iterations as f64;

    println!("[{}] Average: {:.5} ns", name, avg);
    avg
}

#[cfg(feature = "async-com")]
fn main() {
    const ITERATIONS: u64 = 10_000_000; // 10M loops

    println!("Running Rust (kcom) Async Benchmarks ({} iterations)...", ITERATIONS);
    println!("-----------------------------------------------------");

    // Prepare a COM object.
    let raw_obj = MyAsyncImpl::new_com(MyAsyncImpl).unwrap();
    let vtbl = unsafe { *(raw_obj as *mut *const IMyAsyncOpVtbl) };

    // --- Allocation Benchmark ---

    // 1. kcom: async method -> AsyncOperation allocation
    measure_ns("Rust_kcom_AsyncOp_New", ITERATIONS, || {
        let op = unsafe { ((*vtbl).get_status)(raw_obj) };
        if op.is_null() {
            return;
        }
        unsafe {
            let op = ComRc::<AsyncOperationRaw<i32>>::from_raw_unchecked(op);
            black_box(op.as_ptr());
        }
    });

    // 2. Standard Rust: Box::pin(ready)
    measure_ns("Rust_Box_Pin_Ready", ITERATIONS, || {
        let future = core::future::ready(1i32);
        let boxed = Box::pin(future);
        black_box(boxed);
    });

    // --- Dispatch Benchmark ---

    // Prepare a completed async operation.
    let op = unsafe {
        let ptr = ((*vtbl).get_status)(raw_obj);
        if ptr.is_null() {
            return;
        }
        ComRc::<AsyncOperationRaw<i32>>::from_raw_unchecked(ptr)
    };
    let op_ptr = op.as_ptr();

    // 3. VTable call on AsyncOperation::get_status
    measure_ns("Rust_kcom_AsyncOp_GetStatus", ITERATIONS, || {
        unsafe {
            let mut status = AsyncStatus::Started;
            let _ = ((*(*op_ptr).lpVtbl).get_status)(
                op_ptr as *mut core::ffi::c_void,
                &mut status,
            );
            black_box(status);
        }
    });

    // 4. Standard Rust: direct method call
    let native = ModernImpl;
    measure_ns("Rust_Native_Call", ITERATIONS, || {
        black_box(native.get_status());
    });

    // Cleanup.
    unsafe {
        ComObject::<MyAsyncImpl, IMyAsyncOpVtbl>::shim_release(raw_obj);
    }
}

#[cfg(not(feature = "async-com"))]
fn main() {
    eprintln!("This benchmark requires the `async-com` feature. Run with `--features async-com`." );
}
